src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java

```diff
@@ -0,0 +1,228 @@
+		FederatedRequest request;
+		if( msg instanceof FederatedRequest )
+			request = (FederatedRequest) msg;
+		else
+			throw new DMLRuntimeException("FederatedWorkerHandler: Received object no instance of `FederatedRequest`.");
+		FederatedResponse response = null;
+		FederatedRequest.FedMethod method = request.getMethod();
+		System.out.println("[Federated Worker] Received command: " + method.name());
+		synchronized (_seq) {
+			switch (method) {
+				case READ:
+					try {
+						// params: filename
+						checkNumParams(request.getNumParams(), 1);
+						String filename = (String) request.getParam(0);
+						MatrixCharacteristics mc = new MatrixCharacteristics();
+						mc.setBlocksize(ConfigurationManager.getBlocksize());
+						MatrixObject mo = new MatrixObject(Types.ValueType.FP64, filename);
+						OutputInfo oi = null;
+						InputInfo ii = null;
+						// read metadata
+						try {
+							String mtdname = DataExpression.getMTDFileName(filename);
+							Path path = new Path(mtdname);
+							try (FileSystem fs = IOUtilFunctions.getFileSystem(mtdname)) {
+								if( fs.exists(path) ) {
+									try (BufferedReader br = new BufferedReader(new InputStreamReader(fs.open(path)))) {
+										JSONObject mtd = JSONHelper.parse(br);
+										mc.setRows(mtd.getLong(DataExpression.READROWPARAM));
+										mc.setCols(mtd.getLong(DataExpression.READCOLPARAM));
+										String format = mtd.getString(DataExpression.FORMAT_TYPE);
+										if( format.equalsIgnoreCase(FORMAT_TYPE_VALUE_TEXT) ) {
+											oi = OutputInfo.TextCellOutputInfo;
+										}
+										else if( format.equalsIgnoreCase(FORMAT_TYPE_VALUE_BINARY) ) {
+											oi = OutputInfo.BinaryBlockOutputInfo;
+										}
+										else if( format.equalsIgnoreCase(FORMAT_TYPE_VALUE_MATRIXMARKET) ) {
+											oi = OutputInfo.MatrixMarketOutputInfo;
+										}
+										else if( format.equalsIgnoreCase(FORMAT_TYPE_VALUE_LIBSVM) ) {
+											oi = OutputInfo.LIBSVMOutputInfo;
+										}
+										else {
+											response = new FederatedResponse(FederatedResponse.Type.ERROR,
+													"Could not figure out correct file format form metadata file");
+											break;
+										}
+										ii = OutputInfo.getMatchingInputInfo(oi);
+									}
+								}
+							}
+						}
+						catch (Exception ex) {
+							throw new DMLRuntimeException(ex);
+						}
+						MetaDataFormat mdf = new MetaDataFormat(mc, oi, ii);
+						mo.setMetaData(mdf);
+						
+						long id = _seq.getNextID();
+						_vars.put(id, mo);
+						response = new FederatedResponse(FederatedResponse.Type.SUCCESS, id);
+					}
+					catch (DMLRuntimeException exception) {
+						response = new FederatedResponse(FederatedResponse.Type.ERROR,
+								ExceptionUtils.getFullStackTrace(exception));
+					}
+					finally {
+						ctx.writeAndFlush(response);
+					}
+					break;
+				case VECMATMULT:
+					try {
```

Comment:
Do we have a test case for this case?

Reply:
No I forgot to remove this before committing, this will be part of another PR